src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -364,7 +363,14 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                                     target.is_example() ||
                                     profile.test;
 
-                is_correct_dep && is_actual_dep
+                // If the dependency is optional, then we're only activating it
+                // if the corresponding feature was activated
+                let activated = !d.is_optional() ||
+                                self.resolve.features(pkg.package_id()).map(|f| {
```

Comment:
ah, i incorrectly parsed the map call as being on an iterator. I think `HashSet` would be slightly cleaner, but it's not as important as I thought.